File: 6_last_digit_of_the_sum_of_fibonacci_numbers/by_learners/fibonacci_sum_last_digit.rb
#!/usr/bin/env ruby
# Author: Jehadur Rahman
# Organization: CyArm
# GitHub: https://github.com/JehadurRE
# Website: jehadurre.me
# Date: 2025-07-10 09:43 (Thu)

# by Andronik Ordian

def fib_sum_last_digit(n)
  n = n % 60  # Pisano period for mod 10
  
  return n if n <= 1
  
  previous = 0
  current = 1
  sum = 1
  
  (0...n - 1).each do |i|
    tmp_previous = previous
    previous = current
    current = (tmp_previous + current) % 10
    sum = (sum + current) % 10
  end
  
  sum
end

if __FILE__ == $0
  n = gets.to_i
  puts "#{fib_sum_last_digit(n)}"
end

File: 6_last_digit_of_the_sum_of_fibonacci_numbers/fibonacci_sum_last_digit.hs
-- Author: Jehadur Rahman
-- Organization: CyArm
-- GitHub: https://github.com/JehadurRE
-- Website: jehadurre.me
-- Date: 2025-07-10 09:43 (Thu)

-- by Kirill Elagin

fibonacci_sum_naive :: Integer -> Int
fibonacci_sum_naive = helper (0, 1, 0)
  where
    helper (_, _, s) 0 = s `mod` 10
    helper (a, b, s) i = helper (b, a + b, s + b) (i - 1)

fibonacci_sum_fast :: Integer -> Int
fibonacci_sum_fast n = helper (0, 1, 1) (n' - 1)
  where
    n' = n `mod` 60  -- Pisano period for mod 10
    helper (_, _, s) 0 = s `mod` 10
    helper (a, b, s) i = helper (b `mod` 10, (a + b) `mod` 10, (s + (b `mod` 10)) `mod` 10) (i - 1)

main :: IO ()
main = do
  [n] <- fmap words getLine
  print $ fibonacci_sum_fast (read n)
